### PR TITLE
fixes #13292 - replacing the turbolinks progress bar with a spinner

### PR DESCRIPTION
**Pull request by unorthodoxgeek**

### CODE REVIEW

File: app/helpers/puppetclasses_and_environments_helper.rb [hunk 1]
```diff
@@ -17,7 +17,7 @@ def import_proxy_select(hash)
 
   def import_proxy_links(hash, classes = nil)
     SmartProxy.with_features("Puppet").map do |proxy|
-      display_link_if_authorized(_("Import from %s") % proxy.name, hash.merge(:proxy => proxy), {:class=>classes, :"data-no-turbolink" => true})
+      display_link_if_authorized(_("Import from %s") % proxy.name, hash.merge(:proxy => proxy), {:class=>classes})
```
Comment (unorthodoxgeek) [inline app/helpers/puppetclasses_and_environments_helper.rb:20]: this change is what caused me to replace the progress bar, but it's seemingly unrelated to the task description, should I move it to a new PR?

**Comment by dLobatog** [comment]
@unorthodoxgeek :+1: from me, it will probably have to be removed or improved when Turbolinks 3 comes out with true pjax and partial replacement. I'd prefer to wait for one or two other acks before merging though, as it's an application-wide change.

Just to save some time to others - this replaces the turbolinks blue bar by a spinner that looks like this when pages are loading. 

screenshot from 2016-01-21 11-29-04

**Comment by stbenjam** [comment]
I like it, the bar is too hard to see

**Comment by ares** [comment]
I like it but I don't like how it looks. Could it be more transparent and the loading icon bigger ideally with `Loading...` text?

**Comment by GregSutcliffe** [comment]
+1 to @ares' comment. I never liked the bar, a spinner seems much better.

**Comment by dLobatog** [comment]
I don't think the loading text is really necessary (the spinner visibly indicates what it is + https://www.patternfly.org/widgets/ recommends text only for inline spinners) - but then again I don't want to get in that kind of discussion.

**Comment by unorthodoxgeek** [comment]
made it larger
screen shot 2016-01-21 at 1 14 35 pm

**Comment by GregSutcliffe** [comment]
:+1: works for me

**Comment by ares** [comment]
ok, it's better, I'd like to use it for few days... it pops up on every request which is a bit unusual

**Comment by dLobatog** [comment]
Sorry to be a pain in the neck but - can you ack this @kybaker?

**Comment by unorthodoxgeek** [comment]
@dLobatog @ohadlevy @domcleal - I've updated the PR to make the new spinner the way we handle loading screens throughout the application, let me know what you guys think. (I know that this should be squashed, but I want the logical commit to stay there until this is ready for merge)

@tbrisker - look at the second commit - it also has a suggestion for how to handle ordering reusable actions in the code - using a foreman namespace and relevant namespaces into which we can add functionality - later on dividing into multiple files. Think foreman.controllers.hostEdit object which will hold all the host#edit functionality, and will use foreman.tools.whatever for application-wide functionality.

**Comment by ehelms** [comment]
Examining this as a user, I feel:

 a) it's unexpected behavior, given the browser has a loading spinner
 b) I feel like a lot of processing is occurring in the backend between page changes
 c) inconsistency - I change screens via menu and I get this big spinner, I click on an object which changes the page entirely and I get no such spinner
 d) overall it makes me feel like the application is slower because as a user I am trained to expect bits and pieces of the HTML to appear which keeps my attention while giant spinner makes me want to go find something to do while it spins

**Comment by waldenraines** [comment]
I'm personally not a fan of either turbolinks or this approach.  If your pages transition as full page loads (as opposed to a true single page webapp) then you ought to just let the page load normally.

The only place a spinner or any other progress indicator makes sense to me is when you are loading data via JS from an API.

**Comment by unorthodoxgeek** [comment]
The problem is that the browser's spinner doesn't work with turbolinks.
That's why we had the invisible bar on top, and that's why we need
something more visible, otherwise the app just hangs without the user
knowing it's doing something.
If you have a better suggestion for design - I'd be glad to change it.

**Comment by ehelms** [comment]
What about a smaller spinner? Something similar to what github uses when it is loading. You could first try it to the right of the Foreman logo in the header

**Comment by unorthodoxgeek** [comment]
Will take a look tomorrow

**Comment by kybaker** [comment]
@dLobatog I agree with @waldenraines . I don't see a reason to block all of the page data while we wait for a standard page load. It is just a distraction that will become annoying over time. One positive feature of the Turbolinks solution is it complements the content of the app rather than pulling the users focus. Also the users will be able to process the loaded information while they wait on the rest. Additionally most desktop browsers have a spinner identifying that the data is still loading on the page which makes ours redundant. If anything I don't see a need for any indicator.

**Comment by unorthodoxgeek** [comment]
@kybaker - you think that no indication is better? because no indication means that the app acts like it's hanging. The problem with turbolinks is that once you click on a link, nothing happens. The browser's spinner doesn't get triggered because the request is AJAX, and the user doesn't know anything happened. I'm all for a different design (I am not a designer, and I understand nothing in this area), but in terms of user experience, what we currently have is very bad...

**Comment by unorthodoxgeek** [comment]
updated - now the spinner doesn't block the view, only shows small spinner on the right side of the foreman logo
screen shot 2016-01-26 at 10 41 08 am

**Comment by unorthodoxgeek** [comment]
[test]

**Comment by ehelms** [comment]
I am liking this much better, could you try lightening up the color? Either closer to a white spinner or a brighter blue that contrasts the darker blue background.

**Comment by dLobatog** [comment]
There's a spinner-inverse class [in patternfly](https://www.patternfly.org/widgets/#spinner) for these cases, I think it works very well with our background:

screenshot from 2016-01-26 14-28-49

Sorry to be a pest, will merge after that

**Comment by unorthodoxgeek** [comment]
@dLobatog - done.

**Comment by dLobatog** [comment]
@unorthodoxgeek Jenkins has started the countdown to merge.

**Comment by dLobatog** [comment]
Merged as 9a312280d25989fab53c18c1d616bbfa9ee826e6, thanks @unorthodoxgeek!

**Pull request closed without merging**